src/material/core/focus_indicator/_focus-indicator.scss

```diff
@@ -0,0 +1,54 @@
+@import '../theming/theming';
+
+@mixin mat-focus-indicator() {
+  // Focus indicator base styles.
+  .mat-focus-indicator::before {
+    border-radius: 4px;
+    border-width: 2px;
+    border-style: solid;
+    box-sizing: border-box;
+    position: absolute;
+    pointer-events: none;
+    top: 0;
+    right: 0;
+    bottom: 0;
+    left: 0;
+  }
+
+  // Focus indicator offsets.
+  .mat-focus-indicator {
+    &.mat-button-base::before,
+    &.mat-card::before,
+    &.mat-calendar-body-cell::before,
+    &.mat-sort-header-button::before {
+      margin: -4px;
+    }
+
+    &.mat-tab-link::before,
+    &.mat-tab-label::before {
+      margin: 4px 0;
+    }
+  }
+
+  // Render pseudo-element on focus.
+  .mat-focus-indicator {
+    // Checkbox, Option, Slide Toggle
+    .cdk-focused.mat-checkbox &::before,
+    .cdk-focused.mat-radio-button &::before,
+    .cdk-focused.mat-slide-toggle &::before,
```

Comment:
`mat-checkbox`, `mat-radio-button`, `mat-slide-toggle` have visually-hidden inputs. Thus, cannot render focus indicator on them. If there was a way to visually hide an element (but have its dimensions preserved) then that would be a better approach... but I don't think it's possible...

Reply:
Could you encode this context in a comment?

Reply:
I think `position:absolute`, `opacity:0`/`visibility:hidden`/`clip-path:circle(0)`, `pointer-events:none` should do it (@jelbourn would probably know which of those 3 hiding methods is most accessible).

Reply:
`visibility: hidden` won't work because then the input won't receive focus.
`clip-path: circle(0)` isn't cross browser
`opacity: 0` would also make all children (including the focus indicator) have `opacity: 0`. I think this will also be a problem with any of the approaches above...